### ISSUE TITLE
hmr not work with component which not set as a route component

### ISSUE THREAD

**Issue by yukiChow** [wags1999/angular-hmr-lazy-components#4]
I have a module named index.module, then set a router to lazy load it. 

The router component is like a  layout component which contains some other components, for example header, footer. 

The problem is when I used the hmr-module-helper to wrap my index.module, everytime I modify the files in other components( not the route component ), HMR didn't work correctly, it replaced the changed files but page content still stay no change.

Could you tell me is there any solution to work around? thanks!

here is my code.

index.module.ts:
```
import { NgModule, NgModuleRef } from '@angular/core';
import { CommonModule } from '@angular/common';

import { IndexRoutingModule } from './index-routing.module';
import { IndexLayoutComponent } from '../../components/index/common/index-layout/index-layout.component';
import { HomePageComponent } from '../../components/index/home-page/home-page.component';
import { IndexHeaderComponent } from '../../components/index/common/index-header/index-header.component';
import { IndexFooterComponent } from '../../components/index/common/index-footer/index-footer.component';
import { HmrModuleHelper } from 'src/app/hmr-module-helper';

@NgModule({
    declarations: [
        IndexLayoutComponent,
        HomePageComponent,
        IndexHeaderComponent,
        IndexFooterComponent
    ],
    imports: [CommonModule, IndexRoutingModule]
})
export class IndexModule {
    constructor(moduleRef: NgModuleRef<IndexModule>) {
        HmrModuleHelper.enableHmrRouterNgModule(module, moduleRef);
    }
}

HmrModuleHelper.enableHmrNodeModule(module);

```

index-routing.module:
```
import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { IndexLayoutComponent } from 'src/app/components/index/common/index-layout/index-layout.component';
import { HomePageComponent } from 'src/app/components/index/home-page/home-page.component';

const routes: Routes = [
    {
        path: '',
        component: IndexLayoutComponent,
        children: [
            {
                path: 'index',
                component: HomePageComponent
            }
        ]
    }
];

@NgModule({
    imports: [RouterModule.forChild(routes)],
    exports: [RouterModule]
})
export class IndexRoutingModule {}

```

index-layout.component:
```
<app-index-header></app-index-header>
<router-outlet></router-outlet>
<app-index-footer></app-index-footer>
```

Expect:
The page content change as well.

**Comment by wags1999** [comment]
@yukiChow, so you're making a change in one of the components declared in index.module.ts (like  IndexHeaderComponent) and not seeing it reload on the screen?

What is the output of the console when you make a change to the component?  Usually the console will give you some insight into how/if HMR is working.

Do you have a public repo of your code?  Or can you fork this repo and get your example setup?

**Comment by yukiChow** [comment]
thanks for your reply!

I have set a demo to show my problem. Here is the repo: [https://github.com/yukiChow/hmr-helper-demo](url).

You can run it with `npm run dev` which I have already enabled HMR of this app.

After start the app, you can navigate to `localhost:4200/index`, as you can see, there are 3 components: index-header, index-footer, router-outlet(homepage). 

Anything change to the homepage and index-layout will show in page, but anything change to the index-header or index-footer won't show in page.

Still the console log the changed files correctly but page content stay no change. 

I use your hmr-helpers to wrap my index.module, but I modify the helper a little. I remove the DynamicComponentsService code and usage of it in hmr-helpers. 

I was wondering is it why my HMR not work correctly? But I think the demo of dynamic components in your repo is too complex for me, I have to set an individual module for every dynamic component?

Anyway thank you so much to answer my question, your solution helps me a lot!

**Comment by wags1999** [comment]
@yukiChow , sorry that I didn't respond sooner!

I took a look at your repo, and I couldn't find the components you're referring to, nor is it setup for HMR.  Did you miss pushing a commit?

It sounds like your project doesn't have Dynamic or lazy component creation at all - so you should be able to just use the out-of-the-box HMR setup described here: https://github.com/angular/angular-cli/wiki/stories-configure-hmr  On non-route components like your describing, it should cause the entire app to reload.

**Comment by yukiChow** [comment]
I'm so sorry about that! I did miss pushing my setup of the demo :(
I've uploaded the demo right now, please check it out if you get any spare time.
Really really sorry about wasting your time!

As you said above, if I did not wrap the non-route components with your hmr-helper. It did reload the whole app with HMR, so that's not what I wanted.

Thanks for the answer!

**Comment by Prinsn** [comment]
> It sounds like your project doesn't have Dynamic or lazy component creation at all - so you should be able to just use the out-of-the-box HMR setup described here: https://github.com/angular/angular-cli/wiki/stories-configure-hmr On non-route components like your describing, it should cause the entire app to reload.

`Documentation below is deprecated and we no longer accept PRs to improve this.`

**Comment by wags1999** [comment]
@Prinsn, yes, I saw that note there as well.  It appears that Angular HMR docs haven't been updated for 7 or 8.  I'm not sure why. See https://github.com/angular/angular/issues/29452

**Comment by Prinsn** [comment]
@yukiChow I used https://gist.github.com/JurajMlich/f51dd079c5e71c4a8db532b9dfde0ab8 to get a basic HMR up and working and the only problem I had with whole app reloading was related to our login guard detecting that you weren't logged in on recompile due to the `oidc client` losing track of the user.  Once I added in a manual session storage set and retrieve, it appeared to prevent full reload of the core module.

I'm here, as this project was posted shortly after that gist, trying to figure out how to get it to work for the _other_ modules, as they do complete reloads, it seems.
